design-proposals/VM-cloning.md

```diff
@@ -0,0 +1,112 @@
+* Disks
+* Network configuration
+* Volumes
+* CPU topology
+
+Also, provide an ability to fine-tune what's being cloned. For example, it is possible to clone or skip
+annotations / labels.  
+
+## Non Goals
+In first implementation iteration the following limitations exist:
+* Only offline cloning will be supported, meaning cloning a VM when it's stopped.
+* Cloned VM would be in the same namespace as the original one.
+
+## Definition of Users
+Everyone who creates a VM might create a VM clone.
+
+## User Stories
+* As a user / admin, I want to be able to create a new VM that is based on an existing VM.
+* As a user / admin, I want to create a clone to a VM using virtctl.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+Design can be inspired by the way VMIs are being migrated. In order to migrate a VMI, a [VirtualMachineInstanceMigration](https://kubevirt.io/user-guide/operations/live_migration/#initiate-live-migration)
+CRD is posted into the cluster. This CRD is very thin and basically triggers a migration with basic parameters.
+
+For VM cloning a similar approach can be taken: a new CRD would be introduced that would provide some basic parameters like:
+* The VM that needs to be cloned
+* A new name for the VM (could be optional - a new name will be generated)
+* New MAC address (optional: if not provided MAC would be erased)
+* Whether to clone or skip labels / annotations etc.
+
+More parameters could be added in the future, like the new namespace for the VM, and more.
+
+## API Examples
+As said above, API can be similar to [VirtualMachineInstanceMigration](https://kubevirt.io/user-guide/operations/live_migration/#initiate-live-migration).
+
+Here's a sketch for the new CRD:
+```yaml
+kind: VirtualMachineClone
+  spec:
+    sourceVirtualMachineName: my-cool-vm
+    targetVirtualMachineName: my-cool-clone   # Optional - a new name can be generated, e.g. my-cool-vm-jfh54b 
+    newMacAddress: my-new-mac-address         # Optional - new mac address can be generated automatically
+    
+    # All of the fields below are optional and defaults to true
+    cloneAnnotations: true
+    cloneLabels: true
+    cloneDisks: true
+    cloneNetwork: true
+```
+
+In the future, we may want to support configuring new namespace for the cloned VM. However, this requires
+copying config-maps / secrets to another namespace which could be a potential risk.
+
+## Scalability
+I don't see any scalability issues.
+
+## Update/Rollback Compatibility
+New API so should not affect updates / rollbacks.
+
+## Implementation details / challenges (in short)
+Most of the implementation can be fairly simple, but there are a couple of things that better be in mind
+
+* Storage:
+    * Things are pretty easy thanks to CDI's [smart-cloning](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/smart-clone.md).
+With smart cloning we can simply reference the existing DVs to smart-clone them. Smart cloning is supported only for
```

Comment:
This implies that CDI needs to be installed in order to use the feature.

Reply:
I don't think this is a problem, unless some other mechanism of cloning volumes is better. We can think of CSI cloning (which is seamlessly integrated with CDI already) as an alternative, but that won't help non csi storage providers.

Reply:
Maybe I should add this into the design proposal as a limitation / requirement in order for VM cloning to work. WDYT?

Reply:
The CSI layer supports volume cloning.  So shouldn't the expectation be that for a storage clone to succeed, the admin must be using a plugin that supports volume cloning?

Reply:
sorry, for the late reply. CDI is not automatically installed with KubeVirt, so if you rely on CDI then it makes it a requirement in order to use this feature.

Reply:
Then why mention CDI at all if the cloning design has nothing to do with it?